activerecord/lib/active_record/load_tree.rb

```diff
@@ -0,0 +1,162 @@
+  # at the time of instance creation. This is used to determine if were and how a record
+  # was loaded.
+  #
+  # parent._load_tree.load_call_stack
+  #
+  # The LoadTree class will also keep track of all records that are loaded from the database.
+  # With this feature you can debug how a record was loaded, where it happened in the call stack
+  # and in the tree hierarchy and what else was loaded at the same time.
+  #
+  # See all Records loaded since the debug mode was turned on:
+  #
+  # AciveRecord::LoadTree.loaded_records
+  #
+  #
+  # It is then possible to iterate through these and inspect the full tree of each record,
+  # allowing you to see what was loaded and how it was loaded and debug full trees of records.
+  #
+  class LoadTree
+    attr_reader :association_name, :record_class_name, :loaded_associations,
+                :load_call_stack, :siblings, :parent
+    attr_writer :loaded_association
+    require "weakref"
+
+    # Mark an ActiveRecord instance/s as loaded.
+    def self.add_loaded_records(records)
+      return loaded_records unless ActiveRecord.dynamic_includes_overfetch_debugging?
+      return loaded_records if records.nil?
+      records = Array(records) unless records.is_a?(Array)
+      loaded_records.concat(records).uniq!
+    end
+
+    # All ActiveRecords instances marked as loaded.
+    def self.loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] ||= []
+    end
+
+    # Reset the state of tracked loaded and used records.
+    def self.clear_loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] = []
+    end
+
+    # Create a Load Tree instance representing the loaded record and its place in the overall
+    # hierarchy of the object tree. Any passed parents or sibilings should also include the load_tree
+    # instance on themselves.
+    #
+    # parent: Instance that is the parent of the loaded record.
+    # siblings: Instances that were loaded at the same time as the loaded record.
+    # association_name: The name of the association that was loaded.
+    def initialize(creator: nil, parent: nil, siblings: [], association_name: nil)
+      raise ArgumentError, "siblings must be an array" unless siblings.is_a?(Array)
+      raise ArgumentError, "Cannot be in an association if it is a root element" if parent.nil? && !association_name.nil?
+      raise ArgumentError, "all siblings must be have a load tree" unless siblings.all? { |r| r.respond_to?(:_load_tree) && r._load_tree.is_a?(LoadTree) }
+      set_parent(parent)
+
+      @record_class_name = creator.class.name
+      @association_name = association_name
+      parent._load_tree.add_loaded_association(association_name) unless parent.nil? || association_name.nil?
+      @loaded_associations = []
+      @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
+      reset_siblings(siblings)
+    end
```

Comment:
From a design perspective I'm thinking this would be cleaner written like this:

```ruby
attr_reader :parent, :siblings, :record_class_name, :association_name, :loaded_association, :load_call_stack
def initialize(creator: nil, parent: nil, siblings: [], association_name: nil)
  @record_class_name = creator.class.name
  @association_name = association_name
  @loaded_associations = []
  @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
  @parent = parent
  @siblings = siblings
end

def set_records
  set_parent
  reset_siblings
end

etc
```

Then the caller would become `ActiveRecord::LoadTree.new(creator: self, siblings: siblings, parent: parent, association_name: association_name).set_records`. The changes here ensure initialize is responsible only for initialization, then the methods below make each of their roles clear. By using the attr_reader we now make the parent/siblings etc accessible so we don't need to either pass the parent/sibling OR access them with @. The goal here is a cleaner interface.

Reply:
Should the error checking still occur in the initialize statement?

Reply:
I think the error checking is overkill to be honest. The interface should inform the type of data sent, and since apps shouldn't be initializing their own LoadTree, that initializer isn't the right place to handle it anyway. Basically the LoadTree initializer should always be sent the right data from the caller (in this case wherever it's initialized in AR).